ws/websocket.go

```diff
@@ -604,21 +644,32 @@ func (client *Client) writePump() {
 				if err != nil {
 					client.error(fmt.Errorf("close failed: %w", err))
 				}
+				// Disconnected by user command. Not calling auto-reconnect.
+				// Passing nil will also not call onDisconnected
+				closure(nil)
 				return
 			}
 			err := conn.WriteMessage(websocket.TextMessage, data)
 			if err != nil {
 				client.error(fmt.Errorf("write failed: %w", err))
+				closure(err)
+				client.handleReconnection()
 				return
 			}
 		case <-ticker.C:
+			// Send periodic ping
 			_ = conn.SetWriteDeadline(time.Now().Add(client.timeoutConfig.WriteWait))
 			if err := conn.WriteMessage(websocket.PingMessage, []byte{}); err != nil {
 				client.error(fmt.Errorf("write failed: %w", err))
+				closure(err)
+				client.handleReconnection()
 				return
 			}
 		case closed, ok := <-client.webSocket.closeSignal:
-			if !ok || closed {
```

Comment:
not quite sure if you're consuming the same signal channel twice here?

Reply:
`closeSignal` is only being consumed once by the `writePump`. The `readPump` is the one sending the signal. It's used to make sure write and read coroutines are closed and cleaned up, before re-connecting.